src/Themes.tsx

```diff
@@ -1,75 +1,160 @@
 import {List} from "immutable";
 import * as React from "react";
-import {branch, compose, lifecycle, renderNothing, withHandlers, withState} from "recompose";
+import {STORY_RENDERED} from "@storybook/core-events";
+import {branch, compose, lifecycle, mapProps, renderNothing, withHandlers, withState} from "recompose";
 import {Theme} from "./types/Theme";
+import {metadataKey} from "./constants";
 
 export interface ThemeProps {
     channel: any;
     api: any;
     active: boolean;
 }
 
+export interface ThemeMetadataParams {
+    themes?: Theme[];
+    singleThemeMessage?: string;
+    showSingleThemeButton?: boolean;
+    buttonAttributes?: string[];
+}
+
+interface ThemeExtraProps {
+    singleThemeMessage?: string;
+    showSingleThemeButton: boolean;
+    buttonAttributes: string[];
+}
+
 interface ThemeState {
-    theme: Theme;
-    setTheme: (theme: Theme) => void;
-    themes: List<Theme>;
-    setThemes: (themes: List<Theme>) => void;
+    stateTheme: Theme;
+    setStateTheme: (theme: Theme) => void;
+    stateThemes: List<Theme>;
+    setStateThemes: (themes: List<Theme>) => void;
 }
 
 interface ThemeHandler {
     onSelectTheme: (theme: Theme) => void;
     onReceiveThemes: (theme: Theme[]) => void;
+    onStoryRender: (storyId: string) => void;
 }
 
-type BaseComponentProps = ThemeProps & ThemeState & ThemeHandler;
+type BaseComponentProps = ThemeProps & ThemeExtraProps & ThemeState & ThemeHandler;
 
-const BaseComponent: React.SFC<BaseComponentProps> = ({onSelectTheme, themes, theme}) => (
-    <div style={RowStyle}>
-        {themes.map((th, i) => {
-            const buttonStyle = th === theme ? SelectedButtonStyle : ButtonStyle;
-            return <div style={buttonStyle} key={i} onClick={() => onSelectTheme(th)}>{th.name}</div>;
-        }).toArray()}
-    </div>
-);
+const BaseComponent: React.SFC<BaseComponentProps> =
+    ({onSelectTheme, stateThemes, stateTheme, singleThemeMessage, showSingleThemeButton, buttonAttributes}) => (
+        <div>
+            {stateThemes.size === 1 && singleThemeMessage && (
+                <div style={MessageStyle}>{singleThemeMessage}</div>
+            )}
+            {(stateThemes.size > 1 || showSingleThemeButton) && (
```

Comment:
Might be nice to break some of this down into smaller functions.

E.g.
```
const showThemeButtons = (themes, force = false) => themes.size > 1 || force;
```

Then you could have 2 separate functions which return the content for either the singleThemeMessage or the theme buttons, and have this become:

```
<div>
  {showThemeButtons(stateThemes, showSingleThemeButton) ? renderMessage() : renderButtons()}
</div>
```

Reply:
As discussed offline, happy to waive this now 👍

Reply:
Cool, thanks :thumbsup: